evm/src/cpu/kernel/asm/rlp/decode.asm

```diff
@@ -0,0 +1,145 @@
+// Note: currently, these methods do not check that RLP input is in canonical
+// form; for example a single byte could be encoded with the length-of-length
+// form. Technically an EVM must perform these checks, but we aren't really
+// concerned with it in our setting. An attacker who corrupted consensus could
+// prove a non-canonical state, but this would just temporarily stall the bridge
+// until a fix was deployed. We are more concerned with preventing any theft of
+// assets.
```

Comment:
For example, an attacker could (after compromising consensus) prove a state where an account nonce of 0 was encoded as `[0x81, 0]`, which full nodes would reject since it's not the canonical encoding `[0]`. I think it's unlikely though since an attacker couldn't gain much, besides causing some confusion. If it did happen, we could either
- "upgrade" the bridge to replace the non-canonical state with a (otherwise identical) canonical state, and add canonical checks in the kernel code so it can't happen again
- just change honest full nodes to stop enforcing canonical RLP

Do you agree with the approach? We could add canonical checks, it would just make the decoding code a bit longer and a bit less efficient.

Reply:
Makes sense!
> just change honest nodes to stop enforcing canonical RLP

Would there be an issue with just doing this from the beginning?

Reply:
Yeah that could work! If it's for our own (new) chain